### ISSUE TITLE
VRAM is very high when training with a router

### ISSUE THREAD

**Issue by ccdv-ai** [UKPLab/sentence-transformers#3441]
Trying to train xml-roberta large model with splade and inference-free splace.
While splade training fits, I always get OOM with inference free splade (even with DDP + gradient checkpointing + frozen embedding).
The loss is SpladeLoss with SparseMultipleNegativesRankingLoss (I use 1 positive and 1 negative for each query).
I train on 48Gb gpus

```python
# 1. 
doc_encoder = MLMTransformer(
    MODEL_NAME, 
    max_seq_length=MAX_SEQ_LENGTH,
    #model_args={"gradient_checkpointing": True}
    )
doc_pooling = SpladePooling(
    word_embedding_dimension=doc_encoder.get_sentence_embedding_dimension(),
    pooling_strategy="max"
)

# 2. 
query_encoder = SparseStaticEmbedding(tokenizer=doc_encoder.tokenizer)

# 3. 
router = Router.for_query_document(
    query_modules=[query_encoder],
    document_modules=[doc_encoder, doc_pooling]
)

# 4. 
model = SparseEncoder(modules=[router])
```

**Comment by tomaarsen** [comment]
Hello!

Thank you for opening this. That indeed sounds like a bug of some kind, as the inference-free modules (SparseStaticEmbedding) should be comparatively light in memory usage. One possibility is that the query component of the model isn't correctly placed on bf16/fp16 leading to bigger memory usage for those embeddings?

Or perhaps the Router isn't correctly propagating the gradient checkpointing to the underlying model? I'll have to investigate to find out why this is happening.

- Tom Aarsen

**Comment by ccdv-ai** [comment]
The gap in VRAM is huge, I dont think it is related to gradient checkpointing.
I can train without problem a 0.5B params model on 8192 sequence length without gradient checkpointing.(no router):

- what is the VRAM cost of computing and storing the output of the SparseStaticEmbedding?
- is the config of the underlying model modified?
- how sequence length affects VRAM?

**Comment by tomaarsen** [comment]
> what is the VRAM cost of computing and storing the output of the SparseStaticEmbedding?

The computation should be rather efficient. Unlike a regular transformers model, there are no big intermediary hidden states. Instead, a 0-vector is created with the size of the output, and 1's are inserted at the positions where we have tokens. Then, an point-wise multiplication is used to convert the 1's to weights.

This computation should be smaller in memory usage than the document path that you described, and the output should be identical in memory usage as the document path.

> is the config of the underlying model modified?

It should not be.

> how sequence length affects VRAM?

Exponentially for the document path, and linear for the query path. For the query (SparseStaticEmbedding) path, it's only linear because the input IDs scale linearly: the produced embedding has the same size. Regular SPLADE models, like all Transformers-based models, are exponential in memory usage and processing speed relative to the sequence length.

- Tom Aarsen

**Comment by arthurbr11** [comment]
Hey @ccdv-ai!

I tried to reproduce the problem you encountered using the same model and setup. For testing, I used an H100 with 80GB of memory (a bit bigger than your memory though the specific GPU shouldn't matter for this comparison as what matter it see if one config use more than the other).

I ran training experiments both with and without inference-free training, keeping everything else identical (dataset, GPU configuration, etc.). I didn't observe any significant memory differences between the two runs.

I want to hilight to you that here you are using  xml-roberta large that is a model having a big vocabulary size (120k) so the mlm logit matrix will be big and that's probably why you have a problem to train it the same way as the model you mention here:
>  0.5B params model on 8192 sequence length 

I'm only supposing as i don't know the vocab size of the second model you where mentioning here (as i supposed it wasn't xlm-roberta large that doesn't have this sequence length). 

I'm attaching the two scripts I used for my inference-free and standard training comparisons (notice that the inference free should have pretty good result, but i didnt manage to make the normal one get good training result). To adapt them for your configuration, you'll likely need to:
- Reduce batch_size by half as you have less memory 
- Adjust other hyperparameters accordingly to have proper result

With these scripts i think you would reach the same conclusion that both approaches have equivalent memory usage.

If you'd like me to reproduce your specific error more precisely i would need more details, like your training scripts and dataset.

This would help me better understand the discrepancy you're experiencing.


Scripts Inference-Free Splade:  

<details>
<summary>Details</summary>


```python
import logging

from datasets import load_dataset

from sentence_transformers import (
    SparseEncoder,
    SparseEncoderModelCardData,
    SparseEncoderTrainer,
    SparseEncoderTrainingArguments,
)
from sentence_transformers.models import Router
from sentence_transformers.sparse_encoder.evaluation import SparseNanoBEIREvaluator
from sentence_transformers.sparse_encoder.losses import SparseMultipleNegativesRankingLoss, SpladeLoss
from sentence_transformers.sparse_encoder.models import SparseStaticEmbedding, MLMTransformer, SpladePooling
from sentence_transformers.training_args import BatchSamplers
import random


logging.basicConfig(format="%(asctime)s - %(message)s", datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO)

# 1. Load a model to finetune with 2. (Optional) model card data
model_name = "FacebookAI/xlm-roberta-large"
mlm_transformer = MLMTransformer(model_name)
splade_pooling = SpladePooling(
    pooling_strategy="max", word_embedding_dimension=mlm_transformer.get_sentence_embedding_dimension()
)
router = Router.for_query_document(
    query_modules=[SparseStaticEmbedding(tokenizer=mlm_transformer.tokenizer, frozen=False)],
    document_modules=[mlm_transformer, splade_pooling],
)

model = SparseEncoder(
    modules=[router],
    model_card_data=SparseEncoderModelCardData(
        language="en",
        license="apache-2.0",
        model_name="Inference-free SPLADE xlm-roberta-large trained on MS MARCO triplets",
    ),
)
print(model.max_seq_length)

# 3. Load a dataset to finetune on
dataset_size = 100_000  # We only use the first 100k samples for training
logging.info("The dataset has not been fully stored as texts on disk yet. We will do this now.")
corpus = load_dataset("sentence-transformers/msmarco", "corpus", split="train")
corpus = dict(zip(corpus["passage_id"], corpus["passage"]))
queries = load_dataset("sentence-transformers/msmarco", "queries", split="train")
queries = dict(zip(queries["query_id"], queries["query"]))
dataset = load_dataset("sentence-transformers/msmarco", "triplets", split="train")
random_indices = random.sample(range(len(dataset)), dataset_size)
dataset = dataset.select(random_indices)


def id_to_text_map(batch):
    return {
        "query": [queries[qid] for qid in batch["query_id"]],
        "positive": [corpus[pid] for pid in batch["positive_id"]],
        "negative": [corpus[pid] for pid in batch["negative_id"]],
    }


dataset = dataset.map(id_to_text_map, batched=True, remove_columns=["query_id", "positive_id", "negative_id"])
dataset = dataset.train_test_split(test_size=10_000)
train_dataset = dataset["train"]
eval_dataset = dataset["test"]
logging.info(train_dataset)

# 4. Define a loss function
loss = SpladeLoss(
    model=model,
    loss=SparseMultipleNegativesRankingLoss(model=model),
    query_regularizer_weight=0,
    document_regularizer_weight=3e-4,
)

# 5. (Optional) Specify training arguments
short_model_name = model_name if "/" not in model_name else model_name.split("/")[-1]
run_name = f"if-splade-{short_model_name}-msmarco-mrl"
args = SparseEncoderTrainingArguments(
    # Required parameter:
    output_dir=f"models/{run_name}",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    learning_rate=2e-5,
    learning_rate_mapping={
        r"SparseStaticEmbedding\.weight": 1e-3
    },  # Set a higher learning rate for the SparseStaticEmbedding module
    fp16=True,  # Set to False if you get an error that your GPU can't run on FP16
    bf16=False,  # Set to True if you have a GPU that supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicate samples in a batch
    router_mapping={
        "query": "query",
        "positive": "document",
        "negative": "document",
    },  # Map the column names to the routes
    # Optional tracking/debugging parameters:
    eval_strategy="steps",
    eval_steps=500,
    save_strategy="steps",
    save_steps=500,
    save_total_limit=2,
    logging_steps=50,
    run_name=run_name,  # Will be used in W&B if `wandb` is installed
    seed=42,
)

# 6. (Optional) Create an evaluator & evaluate the base model
dev_evaluator = SparseNanoBEIREvaluator(
    dataset_names=["msmarco", "nfcorpus", "nq"], batch_size=16, max_active_dims=512
)

# 7. Create a trainer & train
trainer = SparseEncoderTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss=loss,
    evaluator=dev_evaluator,
)
trainer.train()

# 8. Evaluate the model performance again after training
dev_evaluator(model)

# 9. Save the trained model
model.save_pretrained(f"models/{run_name}/final")

# 10. (Optional) Push it to the Hugging Face Hub
model.push_to_hub(run_name)

```

</details>

Scripts Splade: 

<details>
<summary>Details</summary>

```python
import logging

from datasets import load_dataset

from sentence_transformers import (
    SparseEncoder,
    SparseEncoderModelCardData,
    SparseEncoderTrainer,
    SparseEncoderTrainingArguments,
)
from sentence_transformers.models import Router
from sentence_transformers.sparse_encoder.evaluation import SparseNanoBEIREvaluator
from sentence_transformers.sparse_encoder.losses import SparseMultipleNegativesRankingLoss, SpladeLoss
from sentence_transformers.sparse_encoder.models import SparseStaticEmbedding, MLMTransformer, SpladePooling
from sentence_transformers.training_args import BatchSamplers
import random


logging.basicConfig(format="%(asctime)s - %(message)s", datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO)

# 1. Load a model to finetune with 2. (Optional) model card data
model_name = "FacebookAI/xlm-roberta-large"
mlm_transformer = MLMTransformer(model_name)
splade_pooling = SpladePooling(
    pooling_strategy="max", word_embedding_dimension=mlm_transformer.get_sentence_embedding_dimension()
)
router = Router.for_query_document(
    query_modules=[SparseStaticEmbedding(tokenizer=mlm_transformer.tokenizer, frozen=False)],
    document_modules=[mlm_transformer, splade_pooling],
)

model = SparseEncoder(
    modules=[mlm_transformer, splade_pooling],
    model_card_data=SparseEncoderModelCardData(
        language="en",
        license="apache-2.0",
        model_name="SPLADE xlm-roberta-large trained on MS MARCO triplets",
    ),
)
print(model.max_seq_length)

# 3. Load a dataset to finetune on
dataset_size = 100_000  # We only use the first 100k samples for training
logging.info("The dataset has not been fully stored as texts on disk yet. We will do this now.")
corpus = load_dataset("sentence-transformers/msmarco", "corpus", split="train")
corpus = dict(zip(corpus["passage_id"], corpus["passage"]))
queries = load_dataset("sentence-transformers/msmarco", "queries", split="train")
queries = dict(zip(queries["query_id"], queries["query"]))
dataset = load_dataset("sentence-transformers/msmarco", "triplets", split="train")
random_indices = random.sample(range(len(dataset)), dataset_size)
dataset = dataset.select(random_indices)


def id_to_text_map(batch):
    return {
        "query": [queries[qid] for qid in batch["query_id"]],
        "positive": [corpus[pid] for pid in batch["positive_id"]],
        "negative": [corpus[pid] for pid in batch["negative_id"]],
    }


dataset = dataset.map(id_to_text_map, batched=True, remove_columns=["query_id", "positive_id", "negative_id"])
dataset = dataset.train_test_split(test_size=10_000)
train_dataset = dataset["train"]
eval_dataset = dataset["test"]
logging.info(train_dataset)

# 4. Define a loss function
loss = SpladeLoss(
    model=model,
    loss=SparseMultipleNegativesRankingLoss(model=model),
    query_regularizer_weight=1e-3,
    document_regularizer_weight=1e-2,
)

# 5. (Optional) Specify training arguments
short_model_name = model_name if "/" not in model_name else model_name.split("/")[-1]
run_name = f"splade-{short_model_name}-msmarco-mrl"
args = SparseEncoderTrainingArguments(
    # Required parameter:
    output_dir=f"models/{run_name}",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    learning_rate=2e-5,
    fp16=True,  # Set to False if you get an error that your GPU can't run on FP16
    bf16=False,  # Set to True if you have a GPU that supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicate samples in a batch
    # Optional tracking/debugging parameters:
    eval_strategy="steps",
    eval_steps=1000,
    save_strategy="steps",
    save_steps=1000,
    save_total_limit=2,
    logging_steps=50,
    run_name=run_name,  # Will be used in W&B if `wandb` is installed
    seed=42,
)

# 6. (Optional) Create an evaluator & evaluate the base model
dev_evaluator = SparseNanoBEIREvaluator(
    dataset_names=["msmarco", "nfcorpus", "nq"], batch_size=16, max_active_dims=512
)

# 7. Create a trainer & train
trainer = SparseEncoderTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss=loss,
    evaluator=dev_evaluator,
)
trainer.train()

# 8. Evaluate the model performance again after training
dev_evaluator(model)

# 9. Save the trained model
model.save_pretrained(f"models/{run_name}/final")

# 10. (Optional) Push it to the Hugging Face Hub
model.push_to_hub(run_name)

```

</details>

* Arthur BRESNU

**Comment by ccdv-ai** [comment]
Seems like it depends on the architecture.
* Tried to run with `Snowflake/snowflake-arctic-embed-l-v2.0` but it is not a MLM model so memory just spikes with a router.
* `modernbert-large` works in both cases
* `EuroBERT/EuroBERT-610m` didn't work cause it requires custom code and I can't pass the `trust_remote_code` arg

edit: Wondering if there is a way to run a CausalLM model instead of MLM? 
Getting OOM without router and incompatible config with a router (Qwen/Qwen3-0.6B-Base)

**Comment by tomaarsen** [comment]
I've found an issue with the Router module that may (partially) explain the higher memory usage: https://github.com/UKPLab/sentence-transformers/issues/3452#issuecomment-3097492904

It looks like the Router module currently isn't correctly defined for `torch`. I'll work on a PR soon.

- Tom Aarsen
